Build bootloader with ASLR support

<!--
Welcome to the PyInstaller issue tracker! Before creating an issue, please heed the following:

1. This tracker should only be used to report bugs and request features / enhancements to PyInstaller
    - For questions and general support, use the mailing list, see
	  <http://www.pyinstaller.org/support.html#mailing-list>
2. Use the search function before creating a new issue. Duplicates will be closed and directed to
   the original discussion.
3. When making a bug report, make sure you provide all required information. The easier it is for
   maintainers to reproduce, the faster it'll be fixed.
-->

<!-- +++ ONLY TEXT +++ DO NOT POST IMAGES +++ -->

## Building bootloader with this source code is failing in FreeBSD 


I'm trying to building bootloader in FreeBSD machine using this source code.. but getting below error.
I want the bootloader should be PIE feature enabled 

Earlier I had installed pyinsteller 4.5.1 with that I got bootloader as executable not dynamic shared object so trying to rebuild the bootloader
root@FreeBSD113:~/ESEB_444/pyinstaller-4.5.1/build/lib/PyInstaller/bootloader/FreeBSD-64bit # file run
run: ELF 64-bit LSB executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 11.3, FreeBSD-style, stripped
root@FreeBSD113:~/ESEB_444/pyinstaller-4.5.1/build/lib/PyInstaller/bootloader/FreeBSD-64bit #


```
root@FreeBSD113:~/ESEB_444/pyinstaller-develop/bootloader # python ./waf -v all
'all' finished successfully (0.000s)
'distclean' finished successfully (0.002s)
Setting top to                           : /root/ESEB_444/pyinstaller-develop/bootloader
Setting out to                           : /root/ESEB_444/pyinstaller-develop/bootloader/build
Python Version                           : 3.9.5 (default, Dec 10 2021, 05:28:27) [Clang 8.0.0 (tags/RELEASE_800/final 356365)]
Checking for 'clang' (C compiler)        : 06:18:32 runner ['/usr/bin/clang', '-dM', '-E', '-']
/usr/bin/clang
Checking size of pointer                 : 06:18:32 runner ['/usr/bin/clang', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_643ebb47e5c1940fdd2fcedf4da4ce14/testbuild/test.c.1.o']
06:18:32 runner ['/usr/bin/clang', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_2c136d99cea43f1f771199ce3678fb6c/testbuild/test.c.1.o']
06:18:32 runner ['/usr/bin/clang', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_2c136d99cea43f1f771199ce3678fb6c/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
8
Platform                                 : FreeBSD-64bit detected based on compiler
Checking for compiler flags -m64         : 06:18:32 runner ['/usr/bin/clang', '-m64', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_0b1bc35faa17ebd4280de83e4961660a/testbuild/test.c.1.o']
yes
Checking for linker flags -m64           : 06:18:32 runner ['/usr/bin/clang', '-m64', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_c7ded3f1916d71aa30f75cab9bea3be4/testbuild/test.c.1.o']
06:18:32 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_c7ded3f1916d71aa30f75cab9bea3be4/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for library thr                 : 06:18:32 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_3ae1951dcb6bbfe3b4e3c28e5332484a/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_3ae1951dcb6bbfe3b4e3c28e5332484a/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-lthr']
yes
Checking for library m                   : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_8e76ded66dc6c67b4e4c105d8c46e7bb/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_8e76ded66dc6c67b4e4c105d8c46e7bb/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-lm']
yes
Checking for library z                   : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_7e77edb1fc887c7190a733f3b101ebb0/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_7e77edb1fc887c7190a733f3b101ebb0/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-lz']
yes
Checking for library cmocka              : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_58db325315388727d7b0ea1bc154bfae/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_58db325315388727d7b0ea1bc154bfae/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-lcmocka']
not found
Checking for function unsetenv           : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_2fb2d430badbe17c018cf04e913c0d28/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_2fb2d430badbe17c018cf04e913c0d28/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for function mkdtemp            : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_ce499de3920398c11a340f5198c2fee9/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_ce499de3920398c11a340f5198c2fee9/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for function dirname            : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_d7b0c6ae8227ce3423b2af7279b8eb76/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_d7b0c6ae8227ce3423b2af7279b8eb76/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for function basename           : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_a0286471fa1e961f2ee1e9a206c4b65e/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_a0286471fa1e961f2ee1e9a206c4b65e/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for function strndup            : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_caffe7146e07123eb0f7c6dee4f27ab5/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_caffe7146e07123eb0f7c6dee4f27ab5/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for function strnlen            : 06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '../test.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_e7995e1a9c20af6741bff867b4d056e6/testbuild/test.c.1.o']
06:18:33 runner ['/usr/bin/clang', '-m64', 'test.c.1.o', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/.conf_check_e7995e1a9c20af6741bff867b4d056e6/testbuild/testprog', '-Wl,-Bstatic', '-Wl,-Bdynamic']
yes
Checking for program 'strip'             : /usr/bin/strip
Checking for program 'strip'             : /usr/bin/strip
'configure' finished successfully (0.629s)
'make_all' finished successfully (0.006s)
Waf: Entering directory `/root/ESEB_444/pyinstaller-develop/bootloader/build/debug'
Erroneous order constraint 'after'='cxxprogram' on task class 'strip'
Erroneous order constraint 'after'='cxxshlib' on task class 'strip'
Erroneous order constraint 'after'='fcprogram' on task class 'strip'
Erroneous order constraint 'after'='fcshlib' on task class 'strip'
[ 1/16] Compiling src/pyi_path.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_path.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_path.c.1.o']
[ 2/16] Compiling src/pyi_exception_dialog.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_exception_dialog.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_exception_dialog.c.1.o']
[ 3/16] Compiling src/pyi_apple_events.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_apple_events.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_apple_events.c.1.o']
[ 4/16] Compiling src/pyi_win32_utils.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_win32_utils.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_win32_utils.c.1.o']
[ 5/16] Compiling src/pyi_main.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_main.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_main.c.1.o']
[ 6/16] Compiling src/pyi_python.c
06:18:33 runner ['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_python.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_python.c.1.o']
In file included from ../../src/pyi_main.c:42:
../../src/pyi_utils.h:55:1: error: unknown type name 'pid_t'
pid_t pyi_utils_get_child_pid();
^
1 error generated.

Waf: Leaving directory `/root/ESEB_444/pyinstaller-develop/bootloader/build/debug'
Build failed
 -> task in 'OBJECTS' failed with exit status 1:
        {task 34481478448: c pyi_main.c -> pyi_main.c.1.o}
['/usr/bin/clang', '-m64', '-O2', '-Isrc', '-I../../src', '-Iwindows', '-I../../windows', '-Izlib', '-I../../zlib', '-D_REENTRANT', '-D_BSD_SOURCE', '-DHAVE_UNSETENV=1', '-DHAVE_MKDTEMP=1', '-DHAVE_DIRNAME=1', '-DHAVE_BASENAME=1', '-DHAVE_STRNDUP=1', '-DHAVE_STRNLEN=1', '-DLAUNCH_DEBUG', '-DNDEBUG', '../../src/pyi_main.c', '-c', '-o/root/ESEB_444/pyinstaller-develop/bootloader/build/debug/src/pyi_main.c.1.o']
root@FreeBSD113:~/ESEB_444/pyinstaller-develop/bootloader # 

```

--- a/bootloader/src/pyi_utils.h
+++ b/bootloader/src/pyi_utils.h
@@ -21,6 +21,10 @@
 
 #include "pyi_archive.h"
 
+#ifndef _WIN32
+#include <sys/types.h> /* pid_t */
+#endif
+
 // some platforms do not provide strnlen
 #ifndef HAVE_STRNLEN
 size_t strnlen(const char *str, size_t n);

